### DISCUSSION TITLE
【委托】 求佬讲解动态规划二进制优化

### DISCUSSION THREAD

**Discussion by yytzs** [share121/inter-knot#2026] [其他]
### **如题**
![微信截图_20240814093142](https://github.com/user-attachments/assets/469e0e13-17e9-4724-8472-8e8dee52ef26)
# 宝物筛选

## 题目描述

终于，破解了千年的难题。小 FF 找到了王室的宝物室，里面堆满了无数价值连城的宝物。

这下小 FF 可发财了，嘎嘎。但是这里的宝物实在是太多了，小 FF 的采集车似乎装不下那么多宝物。看来小 FF 只能含泪舍弃其中的一部分宝物了。

小 FF 对洞穴里的宝物进行了整理，他发现每样宝物都有一件或者多件。他粗略估算了下每样宝物的价值，之后开始了宝物筛选工作：小 FF 有一个最大载重为 $W$ 的采集车，洞穴里总共有 $n$ 种宝物，每种宝物的价值为 $v_i$，重量为 $w_i$，每种宝物有 $m_i$ 件。小 FF 希望在采集车不超载的前提下，选择一些宝物装进采集车，使得它们的价值和最大。

## 输入格式

第一行为一个整数 $n$ 和 $W$，分别表示宝物种数和采集车的最大载重。

接下来 $n$ 行每行三个整数 $v_i,w_i,m_i$。

## 输出格式

输出仅一个整数，表示在采集车不超载的情况下收集的宝物的最大价值。

### 样例输入 #1

```
4 20
3 9 3
5 9 1
9 4 2
8 1 3
```

### 样例输出 #1

```
47
```

## 提示

对于 $30\%$ 的数据，$n\leq \sum m_i\leq 10^4$，$0\le W\leq 10^3$。

对于 $100\%$ 的数据，$n\leq \sum m_i \leq 10^5$，$0\le W\leq 4\times 10^4$，$1\leq n\le 100$。

**Comment by Myitian** [comment]
> 对于 $30$ 的数据，$n\leq \sum m_i\leq 10^4$，$0\le W\leq 10^3$
>
> 对于 $100$ 的数据，$n\leq \sum m_i \leq 10^5$，$0\le W\leq 4\times 10^4$，$1\leq n\le 100$

百分号和公式没出来啊，百分号似乎要转义符（ $\LaTeX$里要用 `\%` 来表示百分号，似乎在 GitHub 要用 `\\%`），而且要在 $ 前面加个空格的
```
对于 $30\\%$ 的数据， $n\leq \sum m_i\leq 10^4$， $0\le W\leq 10^3$

对于 $100\\%$ 的数据， $n\leq \sum m_i \leq 10^5$， $0\le W\leq 4\times 10^4$， $1\leq n\le 100$
```
对于 $30\\%$ 的数据， $n\leq \sum m_i\leq 10^4$， $0\le W\leq 10^3$

对于 $100\\%$ 的数据， $n\leq \sum m_i \leq 10^5$， $0\le W\leq 4\times 10^4$， $1\leq n\le 100$

**Comment by LING-JI-ZZZ** [comment]
`这委托，本蒟蒻接了。
等等，这不背包动态规划吗？一眼丁真了。
代码如下（反正样例过了）：

`#include <bits/stdc++.h>
using namespace std;

int max_value(int n, int W, vector<tuple<int, int, int>>& items) {
	vector<int> dp(W + 1, 0);
	
	for (auto& [v, w, m] : items) {
		for (int j = W; j >= w; --j) {
			for (int k = 1; k <= m && k * w <= j; ++k) {
				dp[j] = max(dp[j], dp[j - k * w] + k * v);
			}
		}
	}
	
	return dp[W];
}

int main() {
	int n, W;
	cin >> n >> W;
	vector<tuple<int, int, int>> items(n);
	
	for (int i = 0; i < n; ++i) {
		int v, w, m;
		cin >> v >> w >> m;
		items[i] = make_tuple(v, w, m);
	}
	
	int result = max_value(n, W, items);
	cout << result << endl;
	
	return 0;
}
`

**Comment by LING-JI-ZZZ** [comment]
si~上面markdown格式没对
代码如下：
```cpp
#include <bits/stdc++.h>
using namespace std;

int max_value(int n, int W, vector<tuple<int, int, int>>& items) {
	vector<int> dp(W + 1, 0);
	
	for (auto& [v, w, m] : items) {
		for (int j = W; j >= w; --j) {
			for (int k = 1; k <= m && k * w <= j; ++k) {
				dp[j] = max(dp[j], dp[j - k * w] + k * v);
			}
		}
	}
	
	return dp[W];
}

int main() {
	int n, W;
	cin >> n >> W;
	vector<tuple<int, int, int>> items(n);
	
	for (int i = 0; i < n; ++i) {
		int v, w, m;
		cin >> v >> w >> m;
		items[i] = make_tuple(v, w, m);
	}
	
	int result = max_value(n, W, items);
	cout << result << endl;
	
	return 0;
}

```

**Reply by LING-JI-ZZZ** [reply]
坏了，刚去luogu交了一下TLE了4个点，等我改改

**Comment by LING-JI-ZZZ** [comment]
算了用01背包吧~01背包luogu过了
代码如下~：
```cpp
#include<bits/stdc++.h>
using namespace std;

int max_value(int n, int W, vector<tuple<int, int, int>>& items) {
    vector<pair<int, int>> new_items; 
    
    for (auto& [v, w, m] : items) {
        for (int k = 1; k <= m; k *= 2) {
            new_items.emplace_back(k * v, k * w);
            m -= k;
        }
        if (m > 0) {
            new_items.emplace_back(m * v, m * w);
        }
    }
    
    vector<int> dp(W + 1, 0);
    for (auto& [v, w] : new_items) {
        for (int j = W; j >= w; --j) {
            dp[j] = max(dp[j], dp[j - w] + v);
        }
    }
    
    return dp[W];
}

int main() {
    int n, W;
    cin >> n >> W;
    vector<tuple<int, int, int>> items(n);
    
    for (int i = 0; i < n; ++i) {
        int v, w, m;
        cin >> v >> w >> m;
        items[i] = make_tuple(v, w, m);
    }
    
    int result = max_value(n, W, items);
    cout << result << endl;
    
    return 0;
}
```

**Comment by LING-JI-ZZZ** [comment]
啦啦啦，ACcode

**Comment by CocoonedSoul** [comment]
绳网上是真有大佬哇，那我是不是也可以发一些奇奇怪怪的东西了？
